core/arch/aarchxx/mangle.c

```diff
@@ -2922,7 +2954,573 @@ mangle_icache_op(dcontext_t *dcontext, instrlist_t *ilist, instr_t *instr,
+        if (instr_num_dsts(instr) == 2) {
+#ifdef AARCH64
+            PRE(ilist, where,
+                XINST_CREATE_store_pair(
+                    dcontext,
+                    opnd_create_base_disp(dr_reg_stolen, DR_REG_NULL, 0,
+                                          TLS_LDSTEX_VALUE2_SLOT,
+                                          IF_ARM_ELSE(OPSZ_8, OPSZ_16)),
+                    opnd_create_reg(value2_reg), opnd_create_reg(scratch)));
+#else
+            /* Pair store requires consecutive register numbers.
+             * XXX: We could store the 2 values at once.
+             */
+            PRE(ilist, where,
+                instr_create_save_to_tls(dcontext, value2_reg, TLS_LDSTEX_VALUE2_SLOT));
+            PRE(ilist, where,
+                instr_create_save_to_tls(dcontext, scratch, TLS_LDSTEX_SIZE_SLOT));
+#endif
+        } else {
+            PRE(ilist, where,
+                instr_create_save_to_tls(dcontext, scratch, TLS_LDSTEX_SIZE_SLOT));
+        }
+        if (should_restore) {
+            PRE(ilist, where, instr_create_restore_from_tls(dcontext, scratch, slot));
+        }
+    } else {
+        /* Insert a label so subsequent stex mangling knows the ldex was here. */
+        instr_t *marker = INSTR_CREATE_label(dcontext);
+        instr_set_note(marker, (void *)DR_NOTE_LDEX);
+        dr_instr_label_data_t *label_data = instr_get_label_data_area(marker);
+        label_data->data[0] = value_reg;  /* Ruled out dr_reg_stolen above. */
+        label_data->data[1] = value2_reg; /* Ruled out dr_reg_stolen above. */
+        PRE(ilist, where, marker);
+    }
+    /* Replace the exclusive load with a non-exclusive version. */
+    instr_t *ld_nonex = create_ld_from_ldex(dcontext, instr);
+    instrlist_preinsert(ilist, instr, ld_nonex);
+    if (swap_reg != DR_REG_NULL) {
+        if (instr_writes_to_reg(instr, swap_reg, DR_QUERY_INCLUDE_COND_DSTS)) {
+            PRE(ilist, where,
+                instr_create_save_to_tls(dcontext, swap_reg, TLS_REG_STOLEN_SLOT));
+        }
+        if (swap_restore) {
+            PRE(ilist, where,
+                instr_create_restore_from_tls(dcontext, swap_reg, swap_slot));
+        }
+    }
+    instrlist_remove(ilist, instr);
+    instr_destroy(dcontext, instr);
+    return next_instr;
+}
+
+static instr_t *
+mangle_exclusive_store(dcontext_t *dcontext, instrlist_t *ilist, instr_t *instr,
+                       instr_t *next_instr)
+{
+    LOG(THREAD, LOG_INTERP, 4, "Converting exclusive store @%p to compare-and-swap\n",
+        get_app_instr_xl8(instr));
+    RSTATS_INC(num_stex2cas);
+    /* TODO i#1698: Preserve ARM predication and add tests. */
+    reg_id_t reg_orig_ld_val = DR_REG_NULL;
+    reg_id_t reg_orig_ld_val2 = DR_REG_NULL;
+    bool ldex_in_same_block = false;
+    for (instr_t *in = instr_get_prev(instr); in != NULL; in = instr_get_prev(in)) {
+        if (instr_is_exclusive_load(in) || instr_is_exclusive_store(in))
```

Comment:
Not sure I understand the request: do you mean to add such a comment down below in the store mangling routine where it looks for the label?  It doesn't seem to be relevant at this line here.

Reply:
Anywhere in this for-loop should be ok. I suggested here as a reader might wonder why we're breaking from the loop without setting `ldex_in_same_block` even when we've found the ldex. It'd be good to mention that the load mangling routine decides whether to use this optimisation or not, based on various factors (and maybe list them here).

Reply:
> even when we've found the ldex. 

I think there's a misunderstanding here: if we see a ldex, that's *a different ldex*: it means two ldex in a row.  So there is no paired strex for the ldex we're mangling, and thus we do not want to perform this optimization.

There is actually a comment explaining this right above: "Bail on optimized mangling if followed by another load or a clear of the monitor before the store, since both invalidate this load's monitor."

Reply:
> there is no paired strex for the ldex we're mangling

We're mangling a store though; this is `mangle_exclusive_store`.

> I think there's a misunderstanding here: if we see a ldex, that's a different ldex: it means two ldex in a row. 

Here, we're walking back from an `stex`, so the first `ldex` we encounter should be the load paired with this `stex` right? And not a different `ldex`.

Reply:
Looks like I've been looking at the code in mangle_exclusive_load all this time.  OK, for the store case, sure, a note about where the label is makes sense.